File: src/apps/desktop/src/api/i18n_api.rs
//! I18n API

use crate::api::app_state::AppState;
use bitfun_core::service::i18n::{sync_global_i18n_service_locale, LocaleId, LocaleMetadata};
use log::{error, info, warn};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use tauri::State;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LocaleMetadataResponse {
    pub id: String,
    pub name: String,
    #[serde(rename = "englishName")]
    pub english_name: String,
    #[serde(rename = "nativeName")]
    pub native_name: String,
    pub rtl: bool,
}

#[derive(Debug, Deserialize)]
pub struct SetLanguageRequest {
    pub language: String,
}

#[derive(Debug, Deserialize)]
pub struct TranslateRequest {
    pub key: String,
    pub args: Option<Value>,
}

#[tauri::command]
pub async fn i18n_get_current_language(state: State<'_, AppState>) -> Result<String, String> {
    let config_service = &state.config_service;

    match config_service
        .get_config::<String>(Some("app.language"))
        .await
    {
        Ok(language) => Ok(LocaleId::from_str(&language)
            .unwrap_or_default()
            .as_str()
            .to_string()),
        Err(_) => Ok("zh-CN".to_string()),
    }
}

#[tauri::command]
pub async fn i18n_set_language(
    state: State<'_, AppState>,
    _app: tauri::AppHandle,
    request: SetLanguageRequest,
) -> Result<String, String> {
    let Some(locale_id) = LocaleId::from_str(&request.language) else {
        return Err(format!("Unsupported language: {}", request.language));
    };
    let language = locale_id.as_str();

    let config_service = &state.config_service;

    match config_service.set_config("app.language", language).await {
        Ok(_) => {
            info!("Language set to: {}", language);

            // Sync the in-memory I18nService so bot/remote-connect responses
            // use the newly selected language without requiring an app restart.
            match sync_global_i18n_service_locale(locale_id).await {
                Ok(true) => {}
                Ok(false) => {
                    warn!(
                        "Global I18nService not initialized after language change: language={}",
                        language
                    );
                }
                Err(e) => {
                    warn!(
                        "Failed to sync I18nService locale after language change: language={}, error={}",
                        language, e
                    );
                }
            }

            #[cfg(target_os = "macos")]
            {
                let has_workspace = state.workspace_path.read().await.is_some();
                let mode = if has_workspace {
                    crate::macos_menubar::MenubarMode::Workspace
                } else {
                    crate::macos_menubar::MenubarMode::Startup
                };
                let edit_mode = *state.macos_edit_menu_mode.read().await;
                let _ = crate::macos_menubar::set_macos_menubar_with_mode(
                    &_app, language, mode, edit_mode,
                );
            }
            Ok(format!("Language switched to: {}", language))
        }
        Err(e) => {
            error!("Failed to set language: language={}, error={}", language, e);
            Err(format!("Failed to set language: {}", e))
        }
    }
}

#[tauri::command]
pub async fn i18n_get_supported_languages() -> Result<Vec<LocaleMetadataResponse>, String> {
    Ok(LocaleMetadata::all()
        .into_iter()
        .map(|locale| LocaleMetadataResponse {
            id: locale.id.as_str().to_string(),
            name: locale.name,
            english_name: locale.english_name,
            native_name: locale.native_name,
            rtl: locale.rtl,
        })
        .collect())
}

#[tauri::command]
pub async fn i18n_get_config(state: State<'_, AppState>) -> Result<Value, String> {
    let config_service = &state.config_service;

    let current_language = match config_service
        .get_config::<String>(Some("app.language"))
        .await
    {
        Ok(language) => LocaleId::from_str(&language)
            .unwrap_or_default()
            .as_str()
            .to_string(),
        Err(_) => "zh-CN".to_string(),
    };

    Ok(serde_json::json!({
        "currentLanguage": current_language,
        "fallbackLanguage": "en-US",
        "autoDetect": false
    }))
}

#[tauri::command]
pub async fn i18n_set_config(state: State<'_, AppState>, config: Value) -> Result<String, String> {
    let config_service = &state.config_service;

    if let Some(language) = config.get("currentLanguage").and_then(|v| v.as_str()) {
        let Some(locale_id) = LocaleId::from_str(language) else {
            return Err(format!("Unsupported language: {}", language));
        };

        match config_service
            .set_config("app.language", locale_id.as_str())
            .await
        {
            Ok(_) => {
                match sync_global_i18n_service_locale(locale_id).await {
                    Ok(true) => {}
                    Ok(false) => {
                        warn!(
                            "Global I18nService not initialized after i18n config save: language={}",
                            locale_id.as_str()
                        );
                    }
                    Err(e) => {
                        warn!(
                            "Failed to sync I18nService locale after i18n config save: language={}, error={}",
                            locale_id.as_str(),
                            e
                        );
                    }
                }
                Ok("i18n config saved".to_string())
            }
            Err(e) => {
                error!(
                    "Failed to save i18n config: language={}, error={}",
                    language, e
                );
                Err(format!("Failed to save i18n config: {}", e))
            }
        }
    } else {
        Ok("i18n config saved (no language change)".to_string())
    }
}

File: src/web-ui/src/infrastructure/api/service-api/I18nAPI.ts
 

import { invoke } from '@tauri-apps/api/core';
import type { LocaleId, LocaleMetadata, I18nConfig } from '@/infrastructure/i18n/types';
import { getLocaleMetadata } from '@/infrastructure/i18n/presets';
import { createLogger } from '@/shared/utils/logger';

const log = createLogger('I18nAPI');

 
interface LocaleMetadataResponse {
  id: string;
  name: string;
  englishName: string;
  nativeName: string;
  rtl: boolean;
}

 
class I18nAPIClass {
   
  async getCurrentLanguage(): Promise<LocaleId> {
    try {
      const language = await invoke<string>('i18n_get_current_language');
      return language as LocaleId;
    } catch (error) {
      log.warn('Failed to get current language, using default', error);
      return 'zh-CN';
    }
  }

   
  async setLanguage(language: LocaleId): Promise<string> {
    return invoke<string>('i18n_set_language', { 
      request: { language }
    });
  }

   
  async getSupportedLanguages(): Promise<LocaleMetadata[]> {
    const response = await invoke<LocaleMetadataResponse[]>('i18n_get_supported_languages');
    
    return response.map(item => {
      const id = item.id as LocaleId;
      const registryMetadata = getLocaleMetadata(id);

      return {
        ...registryMetadata,
        id,
        name: item.name,
        englishName: item.englishName,
        nativeName: item.nativeName,
        rtl: item.rtl,
        dateFormat: registryMetadata?.dateFormat ?? '',
        numberFormat: registryMetadata?.numberFormat ?? {
          decimal: '.',
          thousands: ',',
        },
        shortName: registryMetadata?.shortName ?? item.id,
        aliases: registryMetadata?.aliases ?? [item.id],
        contentFallbacks: registryMetadata?.contentFallbacks ?? ['en-US'],
        builtin: true,
      };
    });
  }

   
  async getConfig(): Promise<I18nConfig> {
    try {
      const config = await invoke<any>('i18n_get_config');
      return {
        currentLanguage: config.currentLanguage || 'zh-CN',
        fallbackLanguage: config.fallbackLanguage || 'en-US',
        autoDetect: config.autoDetect || false,
        loadedNamespaces: [],
      };
    } catch (error) {
      log.warn('Failed to get i18n config, using defaults', error);
      return {
        currentLanguage: 'zh-CN',
        fallbackLanguage: 'en-US',
        autoDetect: false,
        loadedNamespaces: [],
      };
    }
  }

   
  async setConfig(config: Partial<I18nConfig>): Promise<string> {
    return invoke<string>('i18n_set_config', { config });
  }
}


export const i18nAPI = new I18nAPIClass();
